src/python/pants/backend/docker/goals/package_image.py

```diff
@@ -187,6 +210,82 @@ def get_context_root(self, default_context_root: str) -> str:
         return os.path.normpath(context_root)
 
 
+@dataclass(frozen=True)
+class ImageRefRegistry:
+    registry: DockerRegistryOptions | None
+    repository: str
+    tags: tuple[ImageRefTag, ...]
+
+
+@dataclass(frozen=True)
+class ImageRefTag:
+    template: str
+    formatted: str
+    full_name: str
+
+
+@dataclass(frozen=True)
+class DockerInfoV1:
+    """The format of the .docker-info.json file."""
+
+    version: Literal[1]
+    image_id: str
+    # It'd be good to include the digest here (e.g. to allow 'docker run
+    # registry/repository@digest'), but that is only known after pushing to a V2 registry
+
+    # registry alias or address -> registry (using a dict rather than just a list[registry] for more
+    # convenient look-ups when multiple values exist)
```

Comment:
I see your point, but it also makes it more difficult to work with generically when you don't know the keys of this dictionary. The schema also gets more complicated, so I'd prefer this to be a list of entries rather than a dict.

Reply:
I definitely don't love this idiosyncratic dict approach, but I think it is much easier to work with non-generically, and only slightly harder to work with generically.

For instance, with the demo from the PR description, where an image is tagged with `pants-{pants.hash}` and pushed to `@demo-reg` and I want a script to pull out the name (e.g. for putting into a cloud template), this might look like (sorry for the detail, I'm spelling it out for my benefit too):

Shell (via https://stedolan.github.io/jq/):

```shell
# dict
jq -r '.registries["@demo-reg"].tags["pants-{pants.hash}"].name' docker.docker-info.json

# list
jq -r '.registries[] | select(.alias == "demo-reg") | .tags[] | select(.template == "pants-{pants.hash}") | .name' docker.docker-info.json
```
Shell (via https://github.com/tomnomnom/gron):
```shell
# dict
gron docker.docker-info.json | grep '@demo-reg.*pants-{pants\.hash}.*name' | gron --values

# list: much harder
```

Python:

```python
# dict
data = json.load(open(...))
name = data["registries"]["@demo-reg"]["tags"]["pants-{pants.hash}"]["name"]

# list
data = json.load(open(...))
# (or with some sort of helper)
demo_reg = next(r for r in data["registries"] if r["alias"] == "demo-reg")
pants_tag = next(t for t in demo_reg["tags"] if t["template"] == "pants-{pants.hash}")
name = pants_tag["name"]
```

JS/TS:

```javascript
// dict
const data = JSON.parse(...)
const name = data.registries["@demo-reg"].tags["pants-{pants.hash}"].name

// list 
const data = JSON.parse(...)
const demoReg = data.registries.find(r => r.alias == "demo-reg")
const pantsTag = demoReg.tags.find(t => t.template == "pants-{pants.hash}")
const name = pantsTag.name
```

Handling the dict approach generically seems like it would swap (in Python syntax) `[handle(nested) for nested in parent]` to `{key: handle(nested) for key, nested in parent.items()}` which is annoying, but seems okay.

None of the examples above are catastrophically bad, so I'm happy to switch if you strongly prefer the more obvious/conventional list style.

Reply:
I've updated this PR, and noticed that this... interacts poorly with #17265: the dict approach works best when there's a unique full name for each (registry, tag) pair, but with `use_local_alias = true`, there can be two, e.g. for `demo-reg`, `latest` (from the PR description), there's `example.invalid/prefix/docker:latest` and `demo-reg/prefix/docker:latest`.

As you can see in the code and/or PR description, I've take the rather poor approach of adjusting the tag key to have a suffix of `:local alias`, but that's pretty ugly. I think this is suggesting just taking the convenience hit and doing the list approach.

Reply:
My thoughts are that the keys used for the dict based approach feels arbitrary, so for the general case I still prefer using lists. ;) 

If working with the data in any other context than on the command line, it will be easy enough to build a dict view of the list data with something like `{ item.tag: item for item in list }`, where the `item.tag` key could be whatever makes sense for the application.

Reply:
I've made the switch (and updated the PR description with the example JSON). You get to say "I told you so" hehe 😄

Reply:
oh, I didn't expect it to come up so soon (if at all in this case)! :P